File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max = 0;
        int left = 0;
        unordered_set<char> set;
        for (int right = 0; right < s.size(); right++) {
            while (set.count(s[right])) {
                set.erase(s[left]);
                left++;
            }
            set.insert(s[right]);
            max = std::max(max, right - left + 1);
        }
        return max;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-3.java
class Solution {
    public int trap(int[] heights) {

        int n = heights.length;

        int left = 0;
        int right = n - 1;
        int leftMax = heights[left];
        int rightMax = heights[right];
        int sum = 0;
        while (left < right) {
            if (leftMax < rightMax) { // left is bottle neck
                int water = leftMax - heights[left];
                if (water > 0) sum += water;
                left++;
                leftMax = Math.max(leftMax, heights[left]);
            } else {
                int water = rightMax - heights[right];
                if (water > 0) sum += water;
                right--;
                rightMax = Math.max(rightMax, heights[right]);
            }
        }

        return sum;
    }
}
